File: mtgSearch/src/androidTest/java/com/dbottillo/mtgsearchfree/util/FileHelper.java
package com.dbottillo.mtgsearchfree.util;

import android.content.Context;
import android.content.res.Resources;

import com.dbottillo.mtgsearchfree.model.MTGCard;
import com.dbottillo.mtgsearchfree.model.MTGSet;
import com.dbottillo.mtgsearchfree.model.helper.CreateDBAsyncTask;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.io.StringWriter;
import java.io.Writer;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public final class FileHelper {

    private FileHelper() {

    }

    public static ArrayList<MTGSet> readSetListJSON(Context context) throws JSONException {
        int setList = context.getResources().getIdentifier("set_list", "raw", context.getPackageName());
        String jsonString = loadFile(context, setList);
        JSONArray jsonArray = new JSONArray(jsonString);
        ArrayList<MTGSet> sets = new ArrayList<>();
        for (int i = 0; i < jsonArray.length(); i++) {
            JSONObject setJ = jsonArray.getJSONObject(i);
            try {
                MTGSet set = new MTGSet(i);
                set.setName(setJ.getString("name"));
                set.setCode(setJ.getString("code"));
                sets.add(set);
            } catch (Resources.NotFoundException e) {
                LOG.e("e: " + e.getLocalizedMessage());
            }
        }
        return sets;
    }

    public static ArrayList<MTGCard> readSingleSetFile(MTGSet set, Context context) throws JSONException {
        String jsonSetString = loadFile(context, CreateDBAsyncTask.setToLoad(context, set.getCode()));
        JSONObject jsonCards = new JSONObject(jsonSetString);
        JSONArray cardsJ = jsonCards.getJSONArray("cards");
        ArrayList<MTGCard> cards = new ArrayList<>();

        for (int k = 0; k < cardsJ.length(); k++) {
            JSONObject cardJ = cardsJ.getJSONObject(k);
            cards.add(cardFromJSON(cardJ, set));
        }

        return cards;
    }

    private static String loadFile(Context context, int file) throws Resources.NotFoundException {
        InputStream is = context.getResources().openRawResource(file);

        Writer writer = new StringWriter();
        char[] buffer = new char[1024];
        try {
            Reader reader = new BufferedReader(new InputStreamReader(is, "UTF-8"));
            int n;
            while ((n = reader.read(buffer)) != -1) {
                writer.write(buffer, 0, n);
            }
            reader.close();
            is.close();
        } catch (IOException e) {
            return null;
        }

        return writer.toString();
    }

    private static MTGCard cardFromJSON(JSONObject jsonObject, MTGSet set) throws JSONException {
        MTGCard card = new MTGCard();

        card.setCardName(jsonObject.getString("name"));
        card.setType(jsonObject.getString("type"));
        card.belongsTo(set);

        int multicolor;
        int land;
        int artifact;

        if (jsonObject.has("colors")) {
            JSONArray colorsJ = jsonObject.getJSONArray("colors");
            for (int k = 0; k < colorsJ.length(); k++) {
                String color = colorsJ.getString(k);
                card.addColor(color);
            }

            if (colorsJ.length() > 1) {
                multicolor = 1;
            } else {
                multicolor = 0;
            }
            land = 0;
        } else {
            multicolor = 0;
            land = 1;
        }

        if (jsonObject.has("types")) {
            JSONArray typesJ = jsonObject.getJSONArray("types");
            for (int k = 0; k < typesJ.length(); k++) {
                card.addType(typesJ.getString(k));
            }
        }

        if (jsonObject.getString("type").contains("Artifact")) {
            artifact = 1;
        } else {
            artifact = 0;
        }

        if (jsonObject.has("manaCost")) {
            card.setManaCost(jsonObject.getString("manaCost"));
            land = 0;
        }
        card.setRarity(jsonObject.getString("rarity"));

        if (jsonObject.has("multiverseid")) {
            card.setMultiVerseId(jsonObject.getInt("multiverseid"));
        }

        String power = "";
        if (jsonObject.has("power")) {
            power = jsonObject.getString("power");
        }
        card.setPower(power);

        String toughness = "";
        if (jsonObject.has("toughness")) {
            toughness = jsonObject.getString("toughness");
        }
        card.setToughness(toughness);

        if (jsonObject.has("text")) {
            card.setText(jsonObject.getString("text"));
        }

        int cmc = -1;
        if (jsonObject.has("cmc")) {
            cmc = jsonObject.getInt("cmc");
        }
        card.setCmc(cmc);
        card.setMultiColor(multicolor == 1);
        card.setAsALand(land == 1);
        card.setAsArtifact(artifact == 1);

        if (jsonObject.has("rulings")) {
            JSONArray rulingsJ = jsonObject.getJSONArray("rulings");
            for (int k = 0; k < rulingsJ.length(); k++) {
                JSONObject ruling = rulingsJ.getJSONObject(k);
                card.addRuling(ruling.getString("text"));
            }
        }

        if (jsonObject.has("layout")) {
            card.setLayout(jsonObject.getString("layout"));
        }

        if (jsonObject.has("number")) {
            card.setNumber(jsonObject.getString("number"));
        }

        Gson gson = new Gson();
        Type type = new TypeToken<List<String>>() {
        }.getType();

        if (jsonObject.has("names")) {
            String names = jsonObject.getString("names");
            if (names != null) {
                List<String> strings = gson.fromJson(names, type);
                card.setNames(strings);
            }
        }
        if (jsonObject.has("supertypes")) {
            String supertypes = jsonObject.getString("supertypes");
            if (supertypes != null) {
                List<String> strings = gson.fromJson(supertypes, type);
                card.setSuperTypes(strings);
            }
        }
        if (jsonObject.has("flavor")) {
            card.setFlavor(jsonObject.getString("flavor"));
        }
        if (jsonObject.has("artist")) {
            card.setArtist(jsonObject.getString("artist"));
        }
        if (jsonObject.has("loyalty") && !jsonObject.isNull("loyalty")) {
            card.setLoyalty(jsonObject.getInt("loyalty"));
        }
        if (jsonObject.has("printings")) {
            String printings = jsonObject.getString("printings");
            if (printings != null) {
                List<String> strings = gson.fromJson(printings, type);
                card.setPrintings(strings);
            }
        }
        if (jsonObject.has("originalText")) {
            card.setOriginalText(jsonObject.getString("originalText"));
        }
        if (jsonObject.has("mciNumber")) {
            card.setMciNumber(jsonObject.getString("mciNumber"));
        }
        if (jsonObject.has("colorIdentity")) {
            String colorIdentity = jsonObject.getString("colorIdentity");
            if (colorIdentity != null) {
                List<String> strings = gson.fromJson(colorIdentity, type);
                card.setColorsIdentity(strings);
            }
        }
        return card;
    }
}

File: MTGSearch/src/androidTest/kotlin/com/dbottillo/mtgsearchfree/util/FileHelper.kt
package com.dbottillo.mtgsearchfree.util

import android.content.Context
import android.content.res.Resources

import com.dbottillo.mtgsearchfree.model.MTGCard
import com.dbottillo.mtgsearchfree.model.MTGSet
import com.dbottillo.mtgsearchfree.model.helper.CreateDBAsyncTask
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.io.Reader
import java.io.StringWriter
import java.io.Writer
import java.lang.reflect.Type
import java.util.ArrayList

object FileHelper {

    @Throws(JSONException::class)
    fun readSetListJSON(context: Context): ArrayList<MTGSet> {
        val setList = context.resources.getIdentifier("set_list", "raw", context.packageName)
        val jsonString = loadFile(context, setList)
        val jsonArray = JSONArray(jsonString)
        val sets = ArrayList<MTGSet>()
        for (i in 0 until jsonArray.length()) {
            val setJ = jsonArray.getJSONObject(i)
            try {
                val set = MTGSet(i)
                set.name = setJ.getString("name")
                set.code = setJ.getString("code")
                sets.add(set)
            } catch (e: Resources.NotFoundException) {
                LOG.e("e: " + e.localizedMessage)
            }

        }
        return sets
    }

    @Throws(JSONException::class)
    fun readSingleSetFile(set: MTGSet, context: Context): ArrayList<MTGCard> {
        val jsonSetString = loadFile(context, CreateDBAsyncTask.setToLoad(context, set.code))
        val jsonCards = JSONObject(jsonSetString)
        val cardsJ = jsonCards.getJSONArray("cards")
        val cards = ArrayList<MTGCard>()

        for (k in 0 until cardsJ.length()) {
            val cardJ = cardsJ.getJSONObject(k)
            cards.add(cardFromJSON(cardJ, set))
        }

        return cards
    }

    @Throws(Resources.NotFoundException::class)
    private fun loadFile(context: Context, file: Int): String? {
        val inputStream = context.resources.openRawResource(file)

        val writer = StringWriter()
        try {
            val reader = BufferedReader(InputStreamReader(inputStream, "UTF-8"))
            val iterator = reader.lineSequence().iterator()
            while(iterator.hasNext()) {
                val line = iterator.next()
                writer.write(line)
            }
            reader.close()
            inputStream.close()
        } catch (e: IOException) {
            return null
        }

        return writer.toString()
    }

    @Throws(JSONException::class)
    private fun cardFromJSON(jsonObject: JSONObject, set: MTGSet): MTGCard {
        val card = MTGCard()

        card.setCardName(jsonObject.getString("name"))
        card.type = jsonObject.getString("type")
        card.belongsTo(set)

        val multicolor: Int
        var land: Int
        val artifact: Int

        if (jsonObject.has("colors")) {
            val colorsJ = jsonObject.getJSONArray("colors")
            for (k in 0 until colorsJ.length()) {
                val color = colorsJ.getString(k)
                card.addColor(color)
            }

            if (colorsJ.length() > 1) {
                multicolor = 1
            } else {
                multicolor = 0
            }
            land = 0
        } else {
            multicolor = 0
            land = 1
        }

        if (jsonObject.has("types")) {
            val typesJ = jsonObject.getJSONArray("types")
            for (k in 0 until typesJ.length()) {
                card.addType(typesJ.getString(k))
            }
        }

        if (jsonObject.getString("type").contains("Artifact")) {
            artifact = 1
        } else {
            artifact = 0
        }

        if (jsonObject.has("manaCost")) {
            card.manaCost = jsonObject.getString("manaCost")
            land = 0
        }
        card.rarity = jsonObject.getString("rarity")

        if (jsonObject.has("multiverseid")) {
            card.multiVerseId = jsonObject.getInt("multiverseid")
        }

        var power = ""
        if (jsonObject.has("power")) {
            power = jsonObject.getString("power")
        }
        card.power = power

        var toughness = ""
        if (jsonObject.has("toughness")) {
            toughness = jsonObject.getString("toughness")
        }
        card.toughness = toughness

        if (jsonObject.has("text")) {
            card.text = jsonObject.getString("text")
        }

        var cmc = -1
        if (jsonObject.has("cmc")) {
            cmc = jsonObject.getInt("cmc")
        }
        card.cmc = cmc
        card.isMultiColor = multicolor == 1
        card.isLand = land == 1
        card.isArtifact = artifact == 1

        if (jsonObject.has("rulings")) {
            val rulingsJ = jsonObject.getJSONArray("rulings")
            for (k in 0 until rulingsJ.length()) {
                val ruling = rulingsJ.getJSONObject(k)
                card.addRuling(ruling.getString("text"))
            }
        }

        if (jsonObject.has("layout")) {
            card.layout = jsonObject.getString("layout")
        }

        if (jsonObject.has("number")) {
            card.number = jsonObject.getString("number")
        }

        val gson = Gson()
        val type = object : TypeToken<List<String>>() {

        }.type

        if (jsonObject.has("names")) {
            val names = jsonObject.getString("names")
            if (names != null) {
                val strings = gson.fromJson<List<String>>(names, type)
                card.names = strings
            }
        }
        if (jsonObject.has("supertypes")) {
            val supertypes = jsonObject.getString("supertypes")
            if (supertypes != null) {
                val strings = gson.fromJson<List<String>>(supertypes, type)
                card.superTypes = strings
            }
        }
        if (jsonObject.has("flavor")) {
            card.flavor = jsonObject.getString("flavor")
        }
        if (jsonObject.has("artist")) {
            card.artist = jsonObject.getString("artist")
        }
        if (jsonObject.has("loyalty") && !jsonObject.isNull("loyalty")) {
            card.loyalty = jsonObject.getInt("loyalty")
        }
        if (jsonObject.has("printings")) {
            val printings = jsonObject.getString("printings")
            if (printings != null) {
                val strings = gson.fromJson<List<String>>(printings, type)
                card.printings = strings
            }
        }
        if (jsonObject.has("originalText")) {
            card.originalText = jsonObject.getString("originalText")
        }
        if (jsonObject.has("mciNumber")) {
            card.mciNumber = jsonObject.getString("mciNumber")
        }
        if (jsonObject.has("colorIdentity")) {
            val colorIdentity = jsonObject.getString("colorIdentity")
            if (colorIdentity != null) {
                val strings = gson.fromJson<List<String>>(colorIdentity, type)
                card.colorsIdentity = strings
            }
        }
        return card
    }
}